src/commands/bloom_util.rs

```
@@ -0,0 +1,125 @@
+        BloomFilterType {
+            expansion,
+            fp_rate,
+            filters,
+        }
+    }
+
+    pub fn get_memory_usage(&self) -> usize {
+        let mut mem = std::mem::size_of::<BloomFilterType>();
+        for filter in &self.filters {
+            mem += std::mem::size_of::<BloomFilter>()
+                + (filter.bloom.number_of_bits() / 8u64) as usize;
+        }
+        mem
+    }
+
+    pub fn item_exists(&self, item: &[u8]) -> bool {
+        // Check if item exists already.
+        for filter in &self.filters {
+            if filter.bloom.check(item) {
+                return true;
+            }
+        }
+        false
+    }
+
+    pub fn cardinality(&self) -> i64 {
+        let mut cardinality = 0;
+        // Check if item exists already.
+        for filter in &self.filters {
+            cardinality += filter.num_items;
+        }
+        cardinality as i64
+    }
+
+    pub fn capacity(&self) -> i64 {
+        let mut capacity = 0;
+        // Check if item exists already.
+        for filter in &self.filters {
+            capacity += filter.capacity;
+        }
+        capacity as i64
+    }
+
+    pub fn add_item(&mut self, item: &[u8]) -> i64 {
+        // Check if item exists already.
+        if self.item_exists(item) {
+            return 0;
+        }
+        if let Some(filter) = self.filters.last_mut() {
+            if self.expansion == 0 || filter.num_items < filter.capacity {
+                // Add item.
+                filter.bloom.set(item);
+                filter.num_items += 1;
+                return 1;
+            }
+            if filter.num_items >= filter.capacity {
+                let new_capacity = filter.capacity * self.expansion;
+                let mut new_filter = BloomFilter::new(self.fp_rate, new_capacity);
+                // Add item.
+                new_filter.bloom.set(item);
+                new_filter.num_items += 1;
+                self.filters.push(new_filter);
```

Shouldn't we fail the request if the expansion is much higher than we can support?

---

`expansion` really means expansion_rate. I can rename this variable if that would help clarify this.

I don't think there is a limit on the number of sub filters an object can have. But (if we want) we can define a config for this to either silently fail expansion (by allowing a `set`) beyond a limit of X sub filters per object OR explicitly return an error (without a `set`).

---

One other aspect we should handle this checking / rejecting based on memory overhead of every operation that creates a new BloomFilter object (BF.ADD, BF.MADD, BD.RESERVE, BF.INSERT, RDB Load). Before any of these operations, we should probably check the memory usage and reject the operations if there is not sufficient space. We have a mechanism to compute the estimated additional memory overhead per creation.